packages/netlify-cms-backend-gitlab/src/API.js

```diff
@@ -224,58 +260,443 @@ export default class API {
+      return this.uploadAndCommit(entry, files, options);
+    } else {
+      return this.editorialWorkflowGit(entry, files, options);
+    }
+  }
+
+  deleteFile(path, commitMessage, { branch = this.branch, author = this.commitAuthor }) {
+    const commitParams = this.toCommitParams(commitMessage, branch, author);
+    return this.request({
+      method: 'DELETE',
+      url: `${this.repoURL}/repository/files/${encodeURIComponent(path)}`,
+      params: commitParams,
+      // last_commit_id: ???,
+    });
+  }
+
+  // Metadata
 
-  persistFiles = (files, { commitMessage, newEntry }) =>
-    Promise.all(
-      files.map(file =>
-        this.uploadAndCommit(file, { commitMessage, updateFile: newEntry === false }),
+  checkMetadataBranch() {
+    return this.getBranch(CMS_METADATA_BRANCH).catch(error => {
+      if (error instanceof APIError && error.status === 404) {
+        return this.createMetadataBranch();
+      }
+      throw error;
+    });
+  }
+
+  createMetadataBranch() {
+    // The Branches API doesn't support creating orphan branches.
+    // The Commits API doesn't support creating orphan branches.
+    //   https://gitlab.com/gitlab-org/gitlab-ce/issues/2420 only works for empty repositories.
+    throw new Error(`Not Implemented - GitLab API does not support creating orphan branches.'`);
+    // TODO - rudolf - Update README / Installation and Configuration.
+    // Git could be used for creating orphan branches.
+    //   git checkout --orphan _netlify_cms
+    //   git rm --cached *
+    //   echo "# Netlify CMS\n\nThis branch is used by Netlify CMS to store metadata information for files, branches, merge requests.\n" > README.md
+    //   git add README.md
+    //   git commit -m "Initial commit by Netlify CMS"
+    //   git push --set-upstream origin _netlify_cms
+  }
+
+  retrieveMetadata(key) {
+    const metadataKey = `gl.meta.${key}`;
+    const metadataPath = `${key}.json`;
+    return localForage.getItem(metadataKey).then(cached => {
+      if (cached && cached.expires > Date.now()) {
+        return cached.data;
+      }
+      console.log(
+        '%c Checking for MetaData files of %s',
+        'line-height: 30px;text-align: center;font-weight: bold',
+        key,
+      );
+      return this.fetchBlob(metadataPath, CMS_METADATA_BRANCH, true)
+        .then(raw => JSON.parse(raw))
+        .catch(error => {
+          if (error instanceof APIError && error.status === 404) {
+            console.log(
+              '%c %s does not have metadata',
+              'line-height: 30px;text-align: center;font-weight: bold',
+              key,
+            );
+            throw error;
+          }
+          throw error;
+        });
+    });
+  }
+
+  storeMetadata(key, data, { create = false } = {}) {
+    const metadataKey = `gl.meta.${key}`;
+    const metadataPath = `${key}.json`;
+    return this.checkMetadataBranch().then(() =>
+      this.uploadAndCommit(
+        null, // entry is not used
```

Comment:
I assumed that `entry` is not used in `uploadAndCommit`, but we probably should use a separate method for metadata.